rust_src/src/lread.rs

```diff
@@ -138,3 +144,33 @@ pub unsafe fn defvar_per_buffer_offset(
         remacs_sys::emacs_abort();
     }
 }
+
+/// Read one Lisp expression as text from STREAM, return as Lisp object.
+/// If STREAM is nil, use the value of `standard-input' (which see).
+/// STREAM or the value of `standard-input' may be:
+///  a buffer (read from point and advance it)
+///  a marker (read from where it points and advance it)
+///  a function (call it with no arguments for each character,
+///      call it with a char as argument to push a char back)
+///  a string (takes text from string, starting at the beginning)
+///  t (read text line using minibuffer and use it, or read from
+///     standard input in batch mode).
+#[lisp_fn(min = "0")]
+pub fn read(stream: LispObject) -> LispObject {
+    if (stream.is_nil() && unsafe { read_check_stream(globals.Vstandard_input) })
+        || read_check_stream(stream)
+    {
+        let cs = CString::new("Lisp expression: ").unwrap();
```

Comment:
Do you mean my proposed change to the implementation or the whole thing?

Reply:
(Separate from my earlier comment)

Your port using `and` logic does not match the C flow.

The C flow says:

```
if stream is nil set it to Vstandard_input

regardless of where stream came from, now check if it is Qt. If it is use Qread_char instead.

Did we hit a Qread_char case for some reason (our choice above or the user request)? 
return call1 (intern ("read-minibuffer"), build_string ("Lisp expression: "));

Otherwise, return read_internal_start (stream, Qnil, Qnil);
```

Reply:
As I understand it, there are four possibilities for the input `stream`: `Qnil`, `Qt`, `Qread_char`, or something else, and three for `Vstandard_input`: `Qt`, `Qread_char`, or something else.

1. If `stream` is `Qt` or `Qread_char`, then do `call1` etc, else `read_internal_start.
2. If `stream` is `Qnil`, then do `call1` etc if `Vstandard_input` is `Qt` or `Qread_char`, else `read_internal_start`.
3. If `stream` is something else else, do `read_internal_start`.

Is that everything? Is that what the port code does? (Honest questions. Needless to say, the original code is not written in a clear style.)

Reply:
Stepping down to C can take some getting used to.

Think of `stream` as an optional parameter. When `stream`  is `nil` use the value of `Vstandard_input` in its place. This is like a Unix CLI program which reads a file if provided otherwise it reads from stdin.

You do not need to include `Vstandard_input` in your list of possible inputs. All that matters is `stream`. One possible value of `stream` is the value of `Vstandard_input` but that does not change how you process it.

#2 is actually done before #1. Call it step zero if you will. `if stream is empty, set stream to the value of Vstandard_input`.
Your #1 is correct. Where the value of `stream` came from is not important at this point.
#3 is also correct.

```
// Maybe there is a better idiom for this...
let stream = if stream.is_nil() {
    stream = Vstandard_input;
} else {
    stream
};

if stream.eq(Qt) || stream.eq(Qread_char) {
    call1(....)
} else {
    read_internal_start()
}
```

Reply:
According to the C code, the author questions whether `stream` ever has the value `t` or `Qread_char`. This code was last touched in 2013 and before that had not been touched for 28 years. So either it is long forgotten and dusty, or that comment is just one confused contributor and no one else has cleaned it up.....

There is no test I can see for those values.

Reply:
There are two possible outcomes to this function: call `read-minibuffer` or call `read_internal_start`. Let's call these `rm` and `ris` for short. Regardless of how the answer is determined, what circumstances lead to `ris` and what circumstances lead to `rm`? Any functions that lead to the same choices under the same circumstances can be considered equivalent from the point of view of the `read` API. (Whether that API is good is a separate question; we're just talking about preserving behavior at this point.)

There are four possibilities to consider for the function input `stream`: `t`, `nil`, `read-char`, or something else. If `stream` is `t` or `read-char`, then we end up with `rm` straightaway. Further, if `stream` is something else, we end up with `ris`. The only interesting case is when `stream` is `nil`. In that case, we end up with `rm` when `standard-input` is `t` or `read-char` and `ris` otherwise.

```
|-----------+----+-----+-----------+------|
| stream >  |    |     |           |      |
| stdin V   | t  | nil | read-char | else |
|-----------+----+-----+-----------+------|
| t         | rm | rm  | rm        | ris  |
| read-char | rm | rm  | rm        | ris  |
| else      | rm | ris | rm        | ris  |
|-----------+----+-----+-----------+------|
```

The two questions to ask now are: 1) Does that analysis match the behavior of the existing C function? 2) Does that analysis match the behavior of the proposed Rust port?

The existing C function is written in a highly imperative, mutative style. I'm proposing dumping it in the trash because it's hard to analyze and leads to pointless arguments like this one :) The proposed replacement is a lot clearer I think, because it just says what circumstances need to hold to lead to either branch. (That's the hope anyway. Whether it succeeds is again a different question.)

Reply:
Personally, I found your version harder to analyze. Which is what lead to my questions here.

The shadowing of the `stream` variable is more idiomatic Rust in my experience. If we blend our two versions we end up with:

```
let stream = if stream.is_nil() {
    stream = Vstandard_input;
} else {
    stream
};

if read_check_stream(stream) {
    call1(...)
} else {
    read_internal_start(...)
}
```

This is a simpler function to understand. The places for a bug to hide are clear. It is also easier to step through with a debugger. Less comparisons 

In Python I would write `stream = stream or Vstandard_input` but this form of assignment is not supported in Rust to the best of my knowledge so we have the more verbose if/else.